File: iEPG2iCal/main.m
//
//  main.m
//  iEPG2iCal
//
//  Created by SUZUKI, Wataru on 4/29/11.
//  Copyright 2011 sz50.com. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <CalendarStore/CalendarStore.h>
#import "iEPG.h"
#import "TVProgramInfo.h"


void parse (NSData *data);

int main (int argc, const char * argv[]) {
    
    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
    
    NSDictionary *stationNumbers = [NSDictionary dictionaryWithObjectsAndKeys:
                                    @"1", @"ＮＨＫ総合" 
                                    ,@"2", @"ＮＨＫ教育" 
                                    ,@"4", @"日本テレビ" 
                                    ,@"5", @"テレビ朝日" 
                                    ,@"6", @"ＴＢＳテレビ"
                                    ,@"7", @"テレビ東京" 
                                    ,@"8", @"フジテレビ" 
                                    ,@"9", @"ＭＸテレビ" 
                                    , nil];
    
    
    NSMutableArray *filePaths = [NSMutableArray array];
    
    NSString *searchPath = [@"~/Downloads" stringByExpandingTildeInPath];
    NSString *outputName;
    NSArray *outputArray;
    [searchPath completePathIntoString:&outputName 
                         caseSensitive:NO
                      matchesIntoArray:&outputArray
                           filterTypes:[NSArray arrayWithObject:@"tvpi"]];
    for (NSString *path in outputArray) {
        if ([path.pathExtension isCaseInsensitiveLike:@"tvpi"]) {
            [filePaths addObject:path];
        }
    }
    
    if (filePaths == nil || filePaths.count == 0) {
        printf("no iEPG files found");
        exit(1);
    }
    
    CalCalendarStore *calendarStore = [CalCalendarStore defaultCalendarStore];
    CalCalendar *calendar = nil;
    for (calendar in [calendarStore calendars]) {
        if ([calendar.title isEqualToString:@"Check"]) {
            break;
        }
    }
    
    NSMutableArray *programs = [NSMutableArray array];
    for (NSString *path in filePaths) {
        iEPG     *iepg1 = [iEPG iEPGWithContentsOfFile:path];
        [programs addObjectsFromArray:iepg1.programInformations];
    }
    
    for (TVProgramInfo *program in programs) {
        NSString *programStation = program.station;
        programStation = [programStation stringByReplacingOccurrencesOfString:@"SPTV"
                                                                   withString:@"CS"];
        programStation = [programStation stringByReplacingOccurrencesOfString:@"BSDT" 
                                                                   withString:@"BS"];
        if ([stationNumbers objectForKey:programStation]) {
            programStation = [stationNumbers objectForKey:programStation];
        }
        
        NSString *memo = program.memo;
        memo = [memo stringByReplacingOccurrencesOfString:@"<BR>" 
                                               withString:@"\r\n"
                                                  options:NSCaseInsensitiveSearch 
                                                    range:NSMakeRange(0, program.memo.length)];
        
        CalEvent *event = [CalEvent event];
        event.calendar  = calendar;
        event.title     = [NSString stringWithFormat:@"[%@] %@", programStation, program.title];
        event.startDate = program.startDateTime;
        event.endDate   = program.endDateTime;
        event.notes     = memo;
        
        NSError *error = nil;
        [calendarStore saveEvent:event span:CalSpanAllEvents error:&error];
        if (error != nil) {
            NSLog(@"%@", error);
        }
    }
    
    [pool drain];
    return 0;
}

File: iEPG2iCal/main.swift
//
//  main.swift
//  iEPG2iCal 2
//
//  Created by Wataru SUZUKI on 8/3/14.
//  Copyright (c) 2014 sz50.com. All rights reserved.
//

import Foundation
import EventKit

let calendarName      = "TV/Check"
let stationReplaceMap = [
    "ＮＨＫ総合"   : "1",
    "ＮＨＫ教育"   : "2",
    "日本テレビ"   : "4",
    "テレビ朝日"   : "5",
    "ＴＢＳテレビ" : "6",
    "テレビ東京"   : "7",
    "フジテレビ"   : "8",
    "ＭＸテレビ"   : "9",
    "ＮＨＫ衛星第一": "BS101",
    "DFS00400"   : "011",
    "DFS00408"   : "021",
    "DFS00410"   : "041",
    "DFS00428"   : "051",
    "DFS00418"   : "061",
    "DFS00430"   : "071",
    "DFS00420"   : "081",
    "DFS05C38"   : "091",
    "SPTV"       : "CS",
    "BSDT"       : "BS",
    "CSDT"       : "CS"
]

var filePaths:   [String] = []
var outputName            = ""
var outputArray: [String] = []

let searchPath = ("~/Downloads" as NSString).standardizingPath
let n = searchPath.completePath(into: &outputName, caseSensitive: Bool(false), matchesInto: &outputArray, filterTypes: ["tvpi", "tvpid"])

for path in outputArray {
    if (path as NSString).pathExtension.count > 0 {
        filePaths.append(path)
    }
}

if (filePaths.count == 0) {
    print("no iEPG files found")
    exit(1)
}

var eventStore          = EKEventStore()
var authorizationStatus = EKEventStore.authorizationStatus(for: EKEntityType.event)

switch authorizationStatus {
case .authorized:
    break

case .notDetermined:
    eventStore.requestAccess(to: EKEntityType.event, completion: { (granted: Bool, e: Error?) in
        if (granted) {
            print("granted")
        }
        } as EKEventStoreRequestAccessCompletionHandler)

case .restricted:
    print("restricted")

case .denied:
    print("denied")

}

var calendarIdentifier: String? = nil
for c: EKCalendar in eventStore.calendars(for: EKEntityType.event) {
    if c.title == calendarName {
        calendarIdentifier = c.calendarIdentifier
        break
    }
}

let calendar: EKCalendar
if let ci = calendarIdentifier,
    let cal = eventStore.calendar(withIdentifier: ci) {
    calendar = cal
} else {
    print("Calendar was not found")
    exit(1)
}

var iepgs: [iEPG] = []
for path in filePaths {
    iepgs.append(try iEPG(path: path))
}

let converter = Converter(withEventStore: eventStore)
converter.stationReplaceMap        = stationReplaceMap
converter.defaultEventAvailability = EKEventAvailability.free

for iepg in iepgs {
    converter.addTVPrograms(iepg.programInformations)
}

try converter.saveToCalendar(calendar)
